### DISCUSSION TITLE
STM32WL55JC

### DISCUSSION THREAD

**Discussion by tcpipchip** [jgromes/RadioLib#775] [General]
Hi
I am using the RADIOLIB...
working cool, the Receive and transmit examples
https://lom204-cli-wisol.blogspot.com/2023/05/o-objetivo-deste-blog-e-demonstrar-como_30.html

Now i am porting to VISUINO!
My question is if the **LoRa protocol** was implemented on STM32WL55JC ?

I need to use LoRa and define
sf , bw , cr, preamble  and txPower 

![image](https://github.com/jgromes/RadioLib/assets/11488460/00106a22-da8f-4edf-9ad3-64bd19ad96e4)

**Comment by jgromes** [comment]
I'm not sure what you mean - the STM32WL series has a built-in LoRa transceiver based on the SX126x. RadioLib provides the interface to it, so it allows you to configure modulation parameters like spreading factor or the bandwidth.

**Reply by tcpipchip** [reply]
Hi
Yes! I forgot that is a sx126x! Good tip!!!
How to enable to use the LoRa protocol in Arduino code ?
This https://github.com/jgromes/RadioLib/blob/master/examples/SX126x/SX126x_Settings/SX126x_Settings.ino to STM32WL55 ?

Thanks!!!

**Comment by tcpipchip** [comment]
Working
The LSM110A (SX1262) now is receiving data from LOM204 (SX1276) using RADIOLIB
![image](https://github.com/jgromes/RadioLib/assets/11488460/07c263d9-1467-4663-ad60-da596ae2641c)
Btw, do you have a receive and transmit example in one ? (not isolated)

**Reply by tcpipchip** [reply]
Doing that…

**Reply by tcpipchip** [reply]
still not having success on Receiving+Transmitting on the same software...

Please, have you example ?

**Reply by jgromes** [reply]
Did you have a look at the SX126x PingPong example? https://github.com/jgromes/RadioLib/blob/master/examples/SX126x/SX126x_PingPong/SX126x_PingPong.ino

**Comment by tcpipchip** [comment]
![image](https://github.com/jgromes/RadioLib/assets/11488460/7a9c4e46-ee83-4226-9c24-bacfe93cc267)
https://lom204-cli-wisol.blogspot.com/2023/06/lsm110a-recebendo-dados-do-lom204-via.html

**Comment by tcpipchip** [comment]
Now i will test with 1 master and 4 slaves using my protocol, where you define the Address of Slave that the Master must to make a communication, one slave by slave
to slave i will use (to be listener fist time)
#define INITIATING_NODE  true
and to master (to be sender first time)
//#define INITIATING_NODE  true
Right ?

**Reply by jgromes** [reply]
If your code is based on the ping pong example, then you need to define `INITIATING_NODE` on the node that you want to initiate the communication. I don't know your protocol, but usually communication would be initiated by master.

**Reply by tcpipchip** [reply]
Hi, my protocol is very simple, send a JSON with a slave Id, and get back the answer from Id slave.

It´s working nice today with LMIC...now porting all to your RadioLib...

Pollimg is enough to transmit and receive, right ? Or do you suggest interrupt ?

**Reply by jgromes** [reply]
I would recommend using interrupt-driven, unless there is a good reason not to (there usually isn't).

**Reply by tcpipchip** [reply]
well, it´s dedicated to only transmit and receive packages received/transmit in UART
STM32 has enough speed :)

**Comment by tcpipchip** [comment]
in your TO DO, plans to MESH ?

**Reply by jgromes** [reply]
I don't plan to add this directly to RadioLib - the goal of the library is to be a driver for transceiver ICs. Yes, there are some protocols/ham radio digital modes implemented, but all of them are pre-existing, I don't plan to create any "RadioLibMesh" protocol.

**Comment by tcpipchip** [comment]
@jgromes 
One questions! Depending on antenna positions, I am not getting communication, but they are near. Is that a Tx power value that is too big ? I am using antena with 9db

how can I change the Tx power value ?

**Reply by jgromes** [reply]
There a couple of things to check:

1. Make sure your RF switch configuration matches your hardware. The STM32WL uses a bit unusual RF switch with 3 control pins and a low power/high power Tx mode. You have to make sure that the pinout as well as the logic is correct.
2. Check your antenna frequency - if you're trying to use antenna tuned to different frequency than you are transmitting, than most of that transmitted power will be returning back to your transceiver, possibly damaging it.

You can set the output power in dBm by calling `radio.setOutputPower(power)` - see the documentation: https://jgromes.github.io/RadioLib/class_s_t_m32_w_lx.html#a9afbde1c466907598974cc3923784c5b

**Comment by tcpipchip** [comment]
good morning

1. yes, it´s correct, it´s the same of SDK
 case RBI_SWITCH_OFF:
    {
		#if LSM_RF_CHANNEL_SWITCH
		/* Turn off switch */
		HAL_GPIO_WritePin(RF_SW_CTRL1_GPIO_PORT, RF_SW_CTRL1_PIN, GPIO_PIN_RESET);
		HAL_GPIO_WritePin(RF_SW_CTRL2_GPIO_PORT, RF_SW_CTRL2_PIN, GPIO_PIN_RESET);
		#else   /* LSM_RF_CHANNEL_SWITCH */
		/* Turn off switch */
		**HAL_GPIO_WritePin(RF_SW_CTRL3_GPIO_PORT, RF_SW_CTRL3_PIN, GPIO_PIN_RESET);
		HAL_GPIO_WritePin(RF_SW_CTRL1_GPIO_PORT, RF_SW_CTRL1_PIN, GPIO_PIN_RESET);
		HAL_GPIO_WritePin(RF_SW_CTRL2_GPIO_PORT, RF_SW_CTRL2_PIN, GPIO_PIN_RESET);**
		#endif   /* LSM_RF_CHANNEL_SWITCH */
		break;
    }
    case RBI_SWITCH_RX:
    {
		#if LSM_RF_CHANNEL_SWITCH
		/*Turns On in Rx Mode the RF Switch */
		HAL_GPIO_WritePin(RF_SW_CTRL1_GPIO_PORT, RF_SW_CTRL1_PIN, GPIO_PIN_SET);
		HAL_GPIO_WritePin(RF_SW_CTRL2_GPIO_PORT, RF_SW_CTRL2_PIN, GPIO_PIN_RESET);
		#else   /* LSM_RF_CHANNEL_SWITCH */
		/*Turns On in Rx Mode the RF Switch */
		**HAL_GPIO_WritePin(RF_SW_CTRL3_GPIO_PORT, RF_SW_CTRL3_PIN, GPIO_PIN_SET);
		HAL_GPIO_WritePin(RF_SW_CTRL1_GPIO_PORT, RF_SW_CTRL1_PIN, GPIO_PIN_SET);
		HAL_GPIO_WritePin(RF_SW_CTRL2_GPIO_PORT, RF_SW_CTRL2_PIN, GPIO_PIN_RESET);**
		#endif   /* LSM_RF_CHANNEL_SWITCH */
		break;
    }
    case RBI_SWITCH_RFO_LP:
    {
		#if LSM_RF_CHANNEL_SWITCH
		/*Turns On in Tx Low Power the RF Switch */
		HAL_GPIO_WritePin(RF_SW_CTRL1_GPIO_PORT, RF_SW_CTRL1_PIN, GPIO_PIN_SET);
		HAL_GPIO_WritePin(RF_SW_CTRL2_GPIO_PORT, RF_SW_CTRL2_PIN, GPIO_PIN_SET);
		#else   /* LSM_RF_CHANNEL_SWITCH */
		/*Turns On in Tx Low Power the RF Switch */
		**HAL_GPIO_WritePin(RF_SW_CTRL3_GPIO_PORT, RF_SW_CTRL3_PIN, GPIO_PIN_SET);
		HAL_GPIO_WritePin(RF_SW_CTRL1_GPIO_PORT, RF_SW_CTRL1_PIN, GPIO_PIN_SET);
		HAL_GPIO_WritePin(RF_SW_CTRL2_GPIO_PORT, RF_SW_CTRL2_PIN, GPIO_PIN_SET);**
		#endif   /* LSM_RF_CHANNEL_SWITCH */
		break;
    }
    case RBI_SWITCH_RFO_HP:
    {
		#if LSM_RF_CHANNEL_SWITCH
		/*Turns On in Tx High Power the RF Switch */
		HAL_GPIO_WritePin(RF_SW_CTRL1_GPIO_PORT, RF_SW_CTRL1_PIN, GPIO_PIN_SET);
		HAL_GPIO_WritePin(RF_SW_CTRL2_GPIO_PORT, RF_SW_CTRL2_PIN, GPIO_PIN_SET);
		#else   /* LSM_RF_CHANNEL_SWITCH */
		/*Turns On in Tx High Power the RF Switch */
		**HAL_GPIO_WritePin(RF_SW_CTRL3_GPIO_PORT, RF_SW_CTRL3_PIN, GPIO_PIN_SET);
		HAL_GPIO_WritePin(RF_SW_CTRL1_GPIO_PORT, RF_SW_CTRL1_PIN, GPIO_PIN_RESET);
		HAL_GPIO_WritePin(RF_SW_CTRL2_GPIO_PORT, RF_SW_CTRL2_PIN, GPIO_PIN_SET);**
		#endif   /* LSM_RF_CHANNEL_SWITCH */
		break;
    }

2. YES, the antenna is correct in the Developter Kit and my Breakout Board

3. I will do...

**Comment by tcpipchip** [comment]
btw, thanks to all help

**Comment by tcpipchip** [comment]
still tryng to find the solution why is working only if the antenna are near

**Comment by tcpipchip** [comment]
but the problem looks only on transmission...i can detect on receiving packages of other far radios...like the SX1276...

**Comment by tcpipchip** [comment]
![image](https://github.com/jgromes/RadioLib/assets/11488460/d5bdd970-df62-4e74-80cd-b14185a27bea)
near works

**Comment by tcpipchip** [comment]
RECEIVED DATA: ALEXANDRE
[LSM110A] RSSI:	        	       -127.50 dBm
[LSM110A] SNR:	 	       -18.25 dB

**Reply by jgromes** [reply]
If those are the values you are getting from devices close to each other, then the SNR is much lower than it should be, which suggest an issue somewhere in the hardware, specifically the RF chain.

**Comment by tcpipchip** [comment]
investigating...

**Comment by tcpipchip** [comment]
looks that LSM110A original SDK use

RF_CTRL1_GPIO_Port GPIOB 12
RF_CTRL2_GPIO_Port GPIOC 13

```
 case RBI_SWITCH_OFF:
    {
		**HAL_GPIO_WritePin(RF_SW_CTRL1_GPIO_PORT, RF_SW_CTRL1_PIN, GPIO_PIN_RESET);
		HAL_GPIO_WritePin(RF_SW_CTRL2_GPIO_PORT, RF_SW_CTRL2_PIN, GPIO_PIN_RESET);**
		break;
    }
    case RBI_SWITCH_RX:
    {
		**HAL_GPIO_WritePin(RF_SW_CTRL1_GPIO_PORT, RF_SW_CTRL1_PIN, GPIO_PIN_SET);
		HAL_GPIO_WritePin(RF_SW_CTRL2_GPIO_PORT, RF_SW_CTRL2_PIN, GPIO_PIN_RESET);**
		break;
    }
    case RBI_SWITCH_RFO_LP:
    {
		**HAL_GPIO_WritePin(RF_SW_CTRL1_GPIO_PORT, RF_SW_CTRL1_PIN, GPIO_PIN_SET);
		HAL_GPIO_WritePin(RF_SW_CTRL2_GPIO_PORT, RF_SW_CTRL2_PIN, GPIO_PIN_SET);**
		break;
    }
    case RBI_SWITCH_RFO_HP:
    {
		**HAL_GPIO_WritePin(RF_SW_CTRL1_GPIO_PORT, RF_SW_CTRL1_PIN, GPIO_PIN_SET);
		HAL_GPIO_WritePin(RF_SW_CTRL2_GPIO_PORT, RF_SW_CTRL2_PIN, GPIO_PIN_SET);**
		break;
    }
```


i modified on your init

```
// set RF switch configuration for Nucleo WL55JC1
// NOTE: other boards may be different!
static const uint32_t rfswitch_pins[] =
                         {PB12,  PC13,  PC5};
static const Module::RfSwitchMode_t rfswitch_table[] = {
  {STM32WLx::MODE_IDLE,  {LOW,  LOW,  LOW}},
  {STM32WLx::MODE_RX,    {HIGH, LOW, LOW}},
  {STM32WLx::MODE_TX_LP, {HIGH, HIGH, HIGH}},
  {STM32WLx::MODE_TX_HP, {HIGH, HIGH,  HIGH}},
  END_OF_MODE_TABLE,
};
```

Looks that Solved!
You was right about config

**Reply by jgromes** [reply]
As the note says:

```
// set RF switch configuration for Nucleo WL55JC1
// NOTE: other boards may be different!
```

If your board is not Nucleo WL55JC1, you will have to make sure the configuration is correct. Glad you were able to resolve the issue :)

**Reply by tcpipchip** [reply]
yes, radiolib will the official option to implement my modbus lora communicator on LSM110A

**Comment by tcpipchip** [comment]
now SNR = is between 0 and 7db :)

**Comment by tcpipchip** [comment]
Thanks! Testing
![image](https://github.com/jgromes/RadioLib/assets/11488460/b30c7a76-5ac4-48b2-9e4c-971fb84a191f)

**Comment by tcpipchip** [comment]
https://lom204-cli-wisol.blogspot.com/2023/06/lsm110a-fazendo-comunicacao_27.html
RadioLib Action! :)
LoRa-ModBus

**Comment by tcpipchip** [comment]
https://lom204-cli-wisol.blogspot.com/2023/06/lsm110a-fazendo-comunicacao_28.html

**Comment by tcpipchip** [comment]
https://lom204-cli-wisol.blogspot.com/2023/06/lsm110a-fazendo-comunicacao_70.html
Other Blog using RadioLib

**Comment by tcpipchip** [comment]
if you put the STM32WL55 in receive mode (radio.startReceive())

Can i transmit LoRa packages when i want ?

If you transmit a LoRa package, must to put in receive modem again ?

**Reply by jgromes** [reply]
You can transmit, however, two things to be aware of:

1. When transmission is complete, the radio will signal this by an interrupt, so if you already have an interrupt handler for received packet, it will get triggered. See e.g. the SX126x ping pong example.
2. Afterwards, you will need to put it back into receive mode.

**Reply by tcpipchip** [reply]
I had to do some modifications, because i am using Polling, now ping -pong

I send a Id0 id to slave 0, that answer me
I send a Id1 id to slave 1, that answer me
I send a Id2 id to slave 2, that answer me
I send a Id3 id to slave 3, that answer me

all listen to the sent and answer packages :)

```
// check if the previous operation finished
  if(operationDone) {
    // reset flag
    operationDone = false;

    if(transmitFlag) {
      // the previous operation was transmission, listen for response
      // print the result
      if (transmissionState == RADIOLIB_ERR_NONE) {
        // packet was successfully sent
        //Serial.println(F("transmission finished!"));
	//CustomCode Output Status
	StatusTransmit(F("transmission finished!"));

      } else {
        Serial.print(F("failed, code "));
	//CustomCode Output Status
	StatusTransmit(F("transmission finished!"));
	//CustomCode Output Code
	CodeTransmit(transmissionState);	
        //Serial.println(transmissionState);
      }

      // listen for response
      radio.startReceive();
      transmitFlag = false;

    } else {
      // the previous operation was reception
      // print data and send another packet
      String str;
      int state = radio.readData(str);

      if (state == RADIOLIB_ERR_NONE) {
        // packet was successfully received
        //Serial.println(F("[LSM110A] Received packet!"));

        // print data of the packet
        //Serial.print(F("[LSM110A] Data:\t\t"));
        //Serial.println(str);

	//RadioLib CustomCode OUTPUT
	ReceivedPackage(str);

        // print RSSI (Received Signal Strength Indicator)
        //Serial.print(F("[LSM110A] RSSI:\t\t"));
        //Serial.print(radio.getRSSI());
        //Serial.println(F(" dBm"));

	//RadioLib CustomCode OUTPUT
	SendRSSI(radio.getRSSI());

        // print SNR (Signal-to-Noise Ratio)
        //Serial.print(F("[LSM110A] SNR:\t\t"));
        //Serial.print(radio.getSNR());
        //Serial.println(F(" dB"));

	//RadioLib CustomCode OUTPUT
	SendSNR(radio.getSNR());
	
	// force listen 
        radio.startReceive();
        transmitFlag = false;
      }
    }  
  }
  //WDT WAKE UP
  IWatchdog.reload();
```

all are LISTEN

**Reply by tcpipchip** [reply]
![image](https://github.com/jgromes/RadioLib/assets/11488460/ea3d387d-678a-49f6-a5d6-abd427162e82)
testing the master/slaves RADIOLIB

**Comment by tcpipchip** [comment]
Hey, one final question
Our stm32WL55 module has 22db RF OUTPUT
About the config

  int state = radio.begin(Channel, 125.0, 10, 5, 0x34, 2, 20); <----2  must to be 22 ?
  radio.setOutputPower(22); <--- or here must to be 22 ?

Not sure to put in this parameters...to define the maximum RF OUTPUT...

Can you give a tip ?

**Reply by jgromes** [reply]
The position is correct, see the documentation here: https://jgromes.github.io/RadioLib/class_s_t_m32_w_lx.html#a4f2a9b5a72b5238d2014199d91094f84

It doesn't matter if you put it in the begin method or setOutputPower - radio.begin calls radio.setOutputPower internally.

**Reply by tcpipchip** [reply]
Nice, thanks!

And 22 parameter is correct to my module, that has the maximum RFOUT of 22db?

**Reply by jgromes** [reply]
Once again, the answer is in the documentation: https://jgromes.github.io/RadioLib/class_s_t_m32_w_lx.html#a9afbde1c466907598974cc3923784c5b

To get up to +22 dBm, your module must have the HP amplifier available and you must have correctly configured RF switching.

**Reply by tcpipchip** [reply]
OK, i see! Sorry!

**Comment by tcpipchip** [comment]
btw, 
![image](https://github.com/jgromes/RadioLib/assets/11488460/0ed0ac5b-d3a7-48f1-9a49-bf82fe0610b6)
